scripts/datastore_upgrade.py

```diff
@@ -0,0 +1,324 @@
+  Returns:
+    A list of entities with a limit of batch_size.
+  """
+  return datastore.range_query(APP_ENTITY_TABLE, APP_ENTITY_SCHEMA, last_key,
+    "", batch_size, start_inclusive=False)
+
+def process_entity(entity, datastore, ds_distributed):
+  """ Processes an entity by updating it if necessary and removing tombstones.
+  Args:
+    entity: The entity to process.
+    datastore: A reference to the batch datastore interface.
+    ds_distributed: A reference to the distributed datastore.
+  Returns:
+    True on success, False otherwise.
+  """
+  logger.debug("Process entity {}".format(str(entity)))
+  key = entity.keys()[0]
+  one_entity = entity[key][APP_ENTITY_SCHEMA[0]]
+  version = entity[key][APP_ENTITY_SCHEMA[1]]
+  if one_entity == datastore_server.TOMBSTONE:
+    return delete_entity_from_table(key, datastore)
+
+  entity_proto = entity_pb.EntityProto(one_entity)
+  app_id = entity_proto.key().app()
+  validated_entity = ds_distributed.validated_result(app_id, entity)
+
+  if not validated_entity:
+    return True
+
+  if not (one_entity == validated_entity[key][APP_ENTITY_SCHEMA[0]] and version == validated_entity[key][APP_ENTITY_SCHEMA[1]]):
+    return update_entity_in_table(key, validated_entity, datastore)
+  return True
+
+def update_entity_in_table(key, validated_entity, datastore):
+  """ Updates the APP_ENTITY_TABLE with the valid entity.
+  Args:
+    key: A str representing the row key to update in the table.
+    validated_entity: A validated entity which needs to be updated in place of
+    the current entity.
+    datastore: A reference to the batch datastore interface.
+  Returns:
+    True: if updating entity was successful.
+    False: if an exception was encountered while updating the entity.
+  """
+  cell_values_to_update = {}
+  cell_values_to_update[key] = {
+    APP_ENTITY_SCHEMA[0]: validated_entity[key][APP_ENTITY_SCHEMA[0]],
+    APP_ENTITY_SCHEMA[1]: validated_entity[key][APP_ENTITY_SCHEMA[1]]
+  }
+  try:
+    datastore.batch_put_entity(APP_ENTITY_TABLE, [key], APP_ENTITY_SCHEMA, cell_values_to_update)
+  except AppScaleDBConnectionError as connection_error:
+    logger.error("Error updating key {0}: {1}".format(key, connection_error))
+    return False
+  except Exception as ex:
+    logger.exception("Unexpected exception while updating entity: {}".format(ex))
+    return False
+  return True
+
+def delete_entity_from_table(key, datastore):
+  """ Performs a hard delete on the APP_ENTITY_TABLE for the given row key.
+  Args:
+    key: A str representing the row key to delete from the table.
+    datastore: A reference to the batch datastore interface.
+  Returns:
+    True: if deleting the entity was successful.
+    False: if an exception was encountered while deleting the entity.
+  """
+  try:
+    datastore.batch_delete(APP_ENTITY_TABLE, [key])
+  except AppScaleDBConnectionError as connection_error:
+    logger.error("Error deleting key {0}: {1}".format(key, connection_error))
+    return False
```

Comment:
I think it would be better to let the exception propagate here rather than return false. As it is now, the return value is ignored in the validate function, so the deletion will silently fail. It would be better to know there is a failure.